File: Arrays/two_sum/two_sum.cpp
#include <vector>
#include <iostream>
#include <unordered_map>
using namespace std;


vector<int> hashmap_method(vector<int>& nums, int target) {

    unordered_map<int, int> hashmap;
    vector<int> result = { -1, -1};

    for (int i = 0; i < nums.size(); i++) {
        if (hashmap.count(target - nums[i])) {
            result[0] = hashmap[target - nums[i]];
            result[1] = i;
            return result;
        }
        hashmap[nums[i]] = i;
    }
    return result;

}

void stdout_array(vector<int> nums) {
    
    for (int i = 0; i < nums.size(); i++) {
        cout << nums[i] << " ";
    }

}

int main() {
    vector<int> nums = {2,7,11,15};
    vector<int> result;
    int target = 9;

    cout << "test case 1: ";
    stdout_array(nums);
    cout << endl; 
    cout << "target : " << target << endl;
    result = hashmap_method(nums, target);
    cout << "result : ";
    stdout_array(result);
    cout << endl;

    return 0;
}

File: Arrays/two_sum/two_sum.rs

use std::collections::HashMap;

fn hashmap_method(nums: Vec<i32>, target: i32) -> Vec<i32> {
    
    let mut hashmap: HashMap<i32, i32> = HashMap::new();
    let mut result: Vec<i32> = Vec::new();
    let mut i:i32 = 0;
    let mut difference:i32;

    result.push(-1);
    result.push(-1);

    for x in nums.iter() {
        difference = target - x;
        if hashmap.contains_key(&difference) {
            result[0 as usize] = *hashmap.get(&difference).unwrap();
            result[1 as usize] = i;
            return result;
        }
        hashmap.insert(*x, i);
        i += 1;
    }
    return result;
}

fn main() {

    let nums: Vec<i32> = vec![2,7,11,15];
    let target:i32 = 9;
    let result: Vec<i32>;

    result = hashmap_method(nums, target);
    println!("result : {:?}", result)

}